Add setup-foundry-and-write-basic-unit-test tutorial

a-ridley opened the pull request:
**Description**:

This tutorial is the first in a short mini-series of testing hedera smart contracts using foundry testing features. The goal is for this first one in the series to replace the standalone tutorial located [here](https://docs.hedera.com/hedera/tutorials/smart-contracts/how-to-set-up-foundry-to-test-smart-contracts-on-hedera). This new tutorial is essentially a revamp of the tutorial linked.

Introducing a new section dedicated to Foundry under Smart Contracts, designed to house all tutorials related to Foundry. 

* a README.md file which introduces Foundry
* A markdown file that outlines the steps to the first tutorial



**Notes for reviewer**:
* The goal is for this first one in the series to replace the standalone tutorial located [here](https://docs.hedera.com/hedera/tutorials/smart-contracts/how-to-set-up-foundry-to-test-smart-contracts-on-hedera). Therefore, once this is approved, we will need to take down the old one.

theekrystallee commented on tutorials/smart-contracts/foundry/setup-foundry-and-write-basic-unit-test.md:
````diff
@@ -0,0 +1,283 @@
+In the terminal, ensure you are in the root project directory and build the project.
+
+```shell
+forge build
````
I ran into a few errors when trying to `forge build`

**First error:**

```bash
Compiler run failed:
Error (7920): Identifier not found or not unique.
 --> test/TodoList.t.sol:7:26:
  |
7 | contract TodoListTest is Test {
  |               
```

**Solution**: I had to update the import path on line 4 of `ToDoList.t.sol` to:
```sol
import "lib/forge-std/src/Test.sol";
```
The import path in `Test.sol` needed to be updated to:
```sol
import {DSTest} from "lib/forge-std/lib/ds-test/src/test.sol";
```

**Second error**: 
```bash
Compiler run failed:
Error (6160): Wrong argument count for function call: 2 arguments given but expected 1.
  --> test/TodoList.t.sol:23:38:
   |
23 |         uint256 numberOfTodosAfter = todoList.createTodo(
   |                                      ^ (Relevant source part starts here and spans across multiple lines).
```

**Solution**: remove one of the arguments in the `test_createTodo_returnsNumberOfTodosIncrementedByOne` function in `ToDoList.t.sol` 

I just commented out line 24 where `address(this)` on line 24 but not sure if the function was meant to take an address AND a string. The `createTodo` function in `ToDoList.sol` needs to be adjusted if that is the case.
<img width="553" alt="Screenshot 2024-01-19 at 7 39 15 PM" src="https://github.com/hashgraph/hedera-docs/assets/56278409/32ac3d6d-23b5-4b45-8769-bd72c9bc6f1b">

`forge build` ran successfully after I made the above edits to the code.
<img width="490" alt="Screenshot 2024-01-19 at 7 47 01 PM" src="https://github.com/hashgraph/hedera-docs/assets/56278409/3d924fa4-da2c-41c7-a24c-58f766f234be">

a-ridley replied:
Great catch Krystal! 

Line 24 has an unnecessary argument. It is needed for the next tutorial but not this first one. I'll update the code.

As for the import issues, I will need to clarify the steps a bit more because the remappings.txt file should take care of it, and there shouldn't be any changes to the import. I'll run through the setup up part on a brand new laptop to double check.

theekrystallee reviewed (changes requested):
Please review the errors I encountered and modify the code if necessary and resubmit. Other than those errors, looks great!

a-ridley commented:
I added steps to ensure `foundryup` is installed in order to be able to use the `forge` command. I also removed the unnecessary argument that was in the test and causing issues. In addition, I moved the step to open the project in a code editor before creating the remappings.txt file. I re-ran through the tutorial on a laptop that did not have foundry previously set up & was able to run `forge build` without any issues.  Let me know if you still run into the import errors. The remappings file should make it so no changes to the import line should be needed. :)

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/create-a-smart-contract.mdx:
```diff
@@ -53,7 +53,8 @@ With [HIP-904](https://hips.hedera.com/hip/hip-904), contracts now support Admin
 | **Constructor Parameters** | The constructor parameters to pass. |
 | **Memo** | The memo to be associated with this contract. (max 100 bytes) |
-| **Hooks** | Details of [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) to attach immediately after creating this contract. See [Create and Manage Hooks](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks). |
+{/* TODO: Restore link to "Create and Manage Hooks" SDK page once that page is published — see https://hips.hedera.com/hip/hip-1195 */}
```
```suggestion
| **Hooks** | Details of [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) to attach immediately after creating this contract. |
{/* TODO: Restore link to "Create and Manage Hooks" SDK page once that page is published — see https://hips.hedera.com/hip/hip-1195 */}
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/create-a-smart-contract.mdx:
```diff
@@ -198,8 +199,9 @@ Creates a smart contract instance using the file ID of the contract bytecode.
 | `setMaxAutomaticTokenAssociations()` | int | Optional |
-| `addHook(<hook>)` | [HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks) | Optional |
-| `setHooks(<hooks>)` | list\<[HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks)> | Optional |
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
```
```suggestion
| `addHook(<hook>)` | HookCreationDetails | Optional |
| `setHooks(<hooks>)` | list\<HookCreationDetails> | Optional |
{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -28,7 +28,8 @@ A transaction that allows you to modify the smart contract entity state like adm
 | **Automatic Token Associations** | The maximum number of tokens that this contract can be automatically associated with (i.e., receive air-drops from). |
 | **Memo** | The new memo to be associated with this contract. |
-| **Hooks** | Add or remove [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) from the contract. Deletions are processed before creations, enabling atomic hook replacement. See [Create and Manage Hooks](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks). |
+{/* TODO: Restore link to "Create and Manage Hooks" SDK page once that page is published — see https://hips.hedera.com/hip/hip-1195 */}
```
```suggestion
| **Hooks** | Add or remove [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) from the contract. Deletions are processed before creations, enabling atomic hook replacement. |
{/* TODO: Restore link to "Create and Manage Hooks" SDK page once that page is published — see https://hips.hedera.com/hip/hip-1195 */}
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -28,7 +28,8 @@ A transaction that allows you to modify the smart contract entity state like adm
 | **Memo** | The new memo to be associated with this contract. |
-| **Hooks** | Add or remove [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) from the contract. Deletions are processed before creations, enabling atomic hook replacement. See [Create and Manage Hooks](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks). |
+{/* TODO: Restore link to "Create and Manage Hooks" SDK page once that page is published — see https://hips.hedera.com/hip/hip-1195 */}
+| **Hooks** | Add or remove [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) from the contract. Deletions are processed before creations, enabling atomic hook replacement. |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/create-a-smart-contract.mdx:
```diff
@@ -198,8 +199,9 @@ Creates a smart contract instance using the file ID of the contract bytecode.
-| `addHook(<hook>)` | [HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks) | Optional |
-| `setHooks(<hooks>)` | list\<[HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks)> | Optional |
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
+| `addHook(<hook>)` | HookCreationDetails | Optional |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/create-a-smart-contract.mdx:
```diff
@@ -198,8 +199,9 @@ Creates a smart contract instance using the file ID of the contract bytecode.
-| `setHooks(<hooks>)` | list\<[HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks)> | Optional |
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
+| `addHook(<hook>)` | HookCreationDetails | Optional |
+| `setHooks(<hooks>)` | list\<HookCreationDetails> | Optional |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/create-a-smart-contract.mdx:
```diff
@@ -53,7 +53,8 @@ With [HIP-904](https://hips.hedera.com/hip/hip-904), contracts now support Admin
 | **Memo** | The memo to be associated with this contract. (max 100 bytes) |
-| **Hooks** | Details of [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) to attach immediately after creating this contract. See [Create and Manage Hooks](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks). |
+{/* TODO: Restore link to "Create and Manage Hooks" SDK page once that page is published — see https://hips.hedera.com/hip/hip-1195 */}
+| **Hooks** | Details of [Hiero Hooks](/hedera/core-concepts/accounts/hiero-hooks) to attach immediately after creating this contract. |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -45,8 +46,9 @@ A transaction that allows you to modify the smart contract entity state like adm
 | `clearAutoRenewAccountId()` | | Optional |
-| `addHookToCreate(<hook>)` | [HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks) | Optional |
-| `setHooksToCreate(<hooks>)` | list\<[HookCreationDetails](/hedera/sdks-and-apis/sdks/accounts-and-hbar/create-and-manage-hooks)> | Optional |
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
```
```suggestion
| `addHookToCreate(<hook>)` | HookCreationDetails | Optional |
| `setHooksToCreate(<hooks>)` | list\<HookCreationDetails> | Optional |
| `addHookToDelete(<hookId>)` | long | Optional |
| `setHooksToDelete(<hookIds>)` | list\<long> | Optional |
{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -45,8 +46,13 @@ A transaction that allows you to modify the smart contract entity state like adm
+| `addHookToDelete(<hookId>)` | long | Optional |
+| `setHooksToDelete(<hookIds>)` | list\<long> | Optional |
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
+| `addHookToCreate(<hook>)` | HookCreationDetails | Optional |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -45,8 +46,13 @@ A transaction that allows you to modify the smart contract entity state like adm
+| `setHooksToDelete(<hookIds>)` | list\<long> | Optional |
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
+| `addHookToCreate(<hook>)` | HookCreationDetails | Optional |
+| `setHooksToCreate(<hooks>)` | list\<HookCreationDetails> | Optional |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -45,8 +46,13 @@ A transaction that allows you to modify the smart contract entity state like adm
+{/* TODO: Restore links to "Create and Manage Hooks" SDK page (HookCreationDetails) once that page is published. */}
+| `addHookToCreate(<hook>)` | HookCreationDetails | Optional |
+| `setHooksToCreate(<hooks>)` | list\<HookCreationDetails> | Optional |
 | `addHookToDelete(<hookId>)` | long | Optional |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/smart-contracts/update-a-smart-contract.mdx:
```diff
@@ -45,8 +46,13 @@ A transaction that allows you to modify the smart contract entity state like adm
+| `addHookToCreate(<hook>)` | HookCreationDetails | Optional |
+| `setHooksToCreate(<hooks>)` | list\<HookCreationDetails> | Optional |
 | `addHookToDelete(<hookId>)` | long | Optional |
 | `setHooksToDelete(<hookIds>)` | list\<long> | Optional |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```suggestion

```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```diff
@@ -41,7 +41,7 @@ Accounts have a property, `maxAutoAssociations`, and the property's value determ
 | Property Value | Description |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```diff
@@ -41,7 +41,7 @@ Accounts have a property, `maxAutoAssociations`, and the property's value determ
 | Property Value | Description |
 |---|---|
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```diff
@@ -41,7 +41,7 @@ Accounts have a property, `maxAutoAssociations`, and the property's value determ
 | Property Value | Description |
 |---|---|
 | `0` | Automatic token associations or **token airdrops** are not allowed, and the account must be manually associated with a token. This also applies if the value is less than or equal to `usedAutoAssociations`. |
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```diff
@@ -41,7 +41,7 @@ Accounts have a property, `maxAutoAssociations`, and the property's value determ
 |---|---|
 | `0` | Automatic token associations or **token airdrops** are not allowed, and the account must be manually associated with a token. This also applies if the value is less than or equal to `usedAutoAssociations`. |
-| `-1` | Unlimited automatic token associations are allowed, and this is the default for accounts created via [auto account creation](/sdks-and-apis/sdks/accounts-and-hbar/create-an-account) and for accounts that began as hollow accounts and are now complete. Accounts with `-1` can receive new tokens without manually associating them. The sender still pays the `maxAutoAssociations` fee and initia ...
+| `-1` | Unlimited automatic token associations are allowed, and this is the default for accounts created via [auto account creation](/hedera/core-concepts/accounts/auto-account-creation) and for accounts that began as hollow accounts and are now complete. Accounts with `-1` can receive new tokens without manually associating them. The sender still pays the `maxAutoAssociations` fee and initial r ...
```
```suggestion
```

theekrystallee commented on hedera/sdks-and-apis/sdks/accounts-and-hbar/create-an-account.mdx:
```diff
@@ -41,7 +41,7 @@ Accounts have a property, `maxAutoAssociations`, and the property's value determ
 | `0` | Automatic token associations or **token airdrops** are not allowed, and the account must be manually associated with a token. This also applies if the value is less than or equal to `usedAutoAssociations`. |
-| `-1` | Unlimited automatic token associations are allowed, and this is the default for accounts created via [auto account creation](/sdks-and-apis/sdks/accounts-and-hbar/create-an-account) and for accounts that began as hollow accounts and are now complete. Accounts with `-1` can receive new tokens without manually associating them. The sender still pays the `maxAutoAssociations` fee and initia ...
+| `-1` | Unlimited automatic token associations are allowed, and this is the default for accounts created via [auto account creation](/hedera/core-concepts/accounts/auto-account-creation) and for accounts that began as hollow accounts and are now complete. Accounts with `-1` can receive new tokens without manually associating them. The sender still pays the `maxAutoAssociations` fee and initial r ...
 | `> 0` | If the value is a positive number (number greater than 0), the number of automatic token associations an account can have is limited to that number. |
```
```suggestion
```
